### PR TITLE
Use kube-builder for GatewayParameters CRD

### PR DESCRIPTION
**Pull request by yuval-k**
suggestion: use kube-builder for CRD generation.
pros: Kube builder generation is simpler, faster, and has more features than skv2 CRD generation.
cons: need to figure out docs integration

### CODE REVIEW

File: projects/gateway2/api/doc.go [hunk 1]
```diff
@@ -0,0 +1 @@
+package api
```
Comment (jenshu) [inline projects/gateway2/api/doc.go:1]: is this file needed?

File: projects/gateway2/api/v1alpha1/kube.go [hunk 2]
```diff
@@ -0,0 +1,265 @@
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Kubernetes autoscaling configuration.
+type Autoscaling struct {
```
Comment (jenshu) [inline projects/gateway2/api/v1alpha1/kube.go:8]: i guess we can remove autoscaling for now if we don't support it yet

File: projects/gateway2/docs/conf/config.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (jenshu) [inline projects/gateway2/docs/conf/config.json:1]: can you add a readme on what the files in this directory are used for, and how/when to edit them?

File: projects/gateway2/deployer/merge.go [hunk 4]
```diff
@@ -52,57 +50,65 @@ func deepMergeSlices[T any](dst, src []T) []T {
...
 		return src
 	}
 
-	dstKube := dst.Spec.GetKube()
-	srcKube := src.Spec.GetKube()
+	dstKube := dst.Spec.Kube
+	srcKube := src.Spec.Kube
 
-	dstKube.EnvoyContainer = deepMergeEnvoyContainer(dstKube.GetEnvoyContainer(), srcKube.GetEnvoyContainer())
+	dstKube.EnvoyContainer = deepMergeEnvoyContainer(dstKube.EnvoyContainer, srcKube.EnvoyContainer)
```
Comment (jenshu) [inline projects/gateway2/deployer/merge.go:74]: can we use getters for all the fields, to avoid potential nil panics?

File: projects/gateway2/deployer/merge.go [hunk 5]
```diff
@@ -52,57 +50,65 @@ func deepMergeSlices[T any](dst, src []T) []T {
...
-			break
-		}
-		dstWorkload = deepMergeDeploymentWorkloadType(dstWorkload, srcWorkload)
-	default:
-		// TODO(jbohanon) log or something? Shouldn't happen unless a new type is added
-		break
-	}
+	dst.Spec.Kube = deepMergeDeploymentWorkloadType(dstKube, srcKube)
+
+	// TODO: Figure out ONEOF behavior
```
Comment (jenshu) [inline projects/gateway2/deployer/merge.go:95]: is this TODO still needed?

File: projects/gateway2/generate.go [hunk 6]
```diff
@@ -1,69 +1,8 @@
...
 
-	"github.com/solo-io/skv2/codegen"
-	"github.com/solo-io/skv2/codegen/model"
-	"github.com/solo-io/skv2/codegen/skv2_anyvendor"
-	"k8s.io/apimachinery/pkg/runtime/schema"
-)
-
-//go:generate go run ./generate.go
 func main() {
-	log.Println("starting generate for k8s gateway controller")
```
Comment (lgadban) [inline projects/gateway2/generate.go:14]: we should replicate this behavior somehow -- we need logging in the generate process that shows progress etc. for troubleshooting purposes

maybe we `go run` a helper file that then runs the `controller-gen` process etc

File: changelog/v1.17.0-beta36/kubebuilder.yaml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+changelog:
+  - type: NON_USER_FACING
```
Comment (jenshu) [inline changelog/v1.17.0-beta36/kubebuilder.yaml:2]: changelog needs to move to 1.18
and should probably be user facing due to gwp api changes

File: projects/gateway2/api/doc.go (hunk 1 above)
Comment (EItanya) [reply projects/gateway2/api/doc.go:1]: Not sure, @yuval-k

File: install/helm/gloo/crds/gateway.gloo.solo.io_gatewayparameters.yaml [hunk 8]
```diff
@@ -0,0 +1,3851 @@
...
+    singular: gatewayparameters
+  scope: Namespaced
+  versions:
+  - name: v1alpha1
+    schema:
+      openAPIV3Schema:
+        properties:
+          apiVersion:
+            description: |-
+              APIVersion defines the versioned schema of this representation of an object.
```
[the remark this answers is not in the document]
Comment (EItanya) [reply install/helm/gloo/crds/gateway.gloo.solo.io_gatewayparameters.yaml:27]: I honestly have no idea, is it important?

File: projects/gateway2/docs/conf/config.json (hunk 3 above)
Comment (EItanya) [reply projects/gateway2/docs/conf/config.json:1]: This stuff is autogenerated, I don't think we need it

File: projects/gateway2/api/v1alpha1/gateway_parameters_types.go [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:categories=gateway-gloo,shortName=gwp
+// +kubebuilder:subresource:status
+type GatewayParameters struct {
```
[the remark this answers is not in the document]
Comment (EItanya) [reply projects/gateway2/api/v1alpha1/gateway_parameters_types.go:11]: Manually created

File: projects/gateway2/deployer/merge.go (hunk 4 above)
Comment (EItanya) [reply projects/gateway2/deployer/merge.go:74]: I added getters manually where I could, but in general we're not generating the code anymore so there are no more getters to use unless we add them.

File: projects/gateway2/generate.go (hunk 6 above)
Comment (EItanya) [reply projects/gateway2/generate.go:14]: To be honest that seems like a rabbit hole

File: projects/gateway2/deployer/merge.go (hunk 5 above)
Comment (EItanya) [reply projects/gateway2/deployer/merge.go:95]: Yes, we need to figure out how we want to handle `oneof` given that we don't have proto anymore. One way could be via `cel` validation. Gateway API handles it with enums that match to fields. @yuval-k @ilackarms @lgadban  do y'all have input on that?

File: install/helm/gloo/crds/gateway.gloo.solo.io_gatewayparameters.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (sam-heilbron) [reply install/helm/gloo/crds/gateway.gloo.solo.io_gatewayparameters.yaml:27]: What we had seen in the past was the CRDs grew (the Proxy CRD in particular) to a size that was too large for etcD. https://github.com/solo-io/gloo/blob/main/test/kube2e/helm/helm_test.go#L424 was created to identify this in the future. So while I agree that the descriptions expand the size of the CRD, I _think_ that test shows we won't hit size limits with this. If we do, then we may have to solve the description challenge, but for now I don't think it's necessary

File: projects/gateway2/api/external/kubernetes/README.md [hunk 10]
```diff
@@ -1,42 +0,0 @@
-## Kubernetes Protos in Gloo Gateway
```
Comment (sam-heilbron) [inline projects/gateway2/api/external/kubernetes/README.md:1]: It kind of feels like this README is actually fairly useful, but we need to update it with the context from this PR: ie why we use Go types and what implication that has for future work

File: projects/gateway2/api/v1alpha1/gateway_parameters_types.go [hunk 11]
```diff
@@ -0,0 +1,510 @@
+package v1alpha1
```
Comment (sam-heilbron) [inline projects/gateway2/api/v1alpha1/gateway_parameters_types.go:1]: How will we "publish" these types to solo-apis? Does that mechanism need to be updated as well, or do we get that for free?

File: projects/gateway2/api/v1alpha1/kube.go [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+
+import (
+	corev1 "k8s.io/api/core/v1"
+)
+
+// A container image. See https://kubernetes.io/docs/concepts/containers/images
+// for details.
+type Image struct {
+	// The image registry.
+	// +kubebuilder:validation:Optional
```
Comment (sam-heilbron) [inline projects/gateway2/api/v1alpha1/kube.go:11]: Do we want to add this into 1.17? We had held off from adding this to our existing APIs to avoid rushing in additional validation late in the cycle that could impact users. But this is a new API so I can see an argument for being more aggressive with adding it. wdyt?

File: projects/gateway2/api/v1alpha1/kube.go [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+	Digest string `json:"digest,omitempty"`
+	// The image pull policy for the container. See
+	// https://kubernetes.io/docs/concepts/containers/images/#image-pull-policy
+	// for details.
+	// +kubebuilder:validation:Optional
+	PullPolicy corev1.PullPolicy `json:"pull_policy,omitempty"`
+}
+
+func (in *Image) GetRegistry() string {
```
Comment (sam-heilbron) [inline projects/gateway2/api/v1alpha1/kube.go:29]: I wonder if a comment here, or a mention in the README about how we don't have `getters` for free could provie some context to users. I imagine the github.com/saiskee/gettercheck doesn't catch this?

**Review by sam-heilbron — COMMENTED** [review]
looking good!

File: projects/gateway2/api/v1alpha1/gateway_parameters_types.go (hunk 11 above)
Comment (jenshu) [reply projects/gateway2/api/v1alpha1/gateway_parameters_types.go:1]: we currently don't publish this to solo-apis, so that will need to be a separate set of work

File: projects/gateway2/deployer/merge.go (hunk 5 above)
Comment (lgadban) [reply projects/gateway2/deployer/merge.go:95]: Did some poking around, IMO we should follow the K8s convention
i.e. use a type field that is effectively a string enum

Looking at GwApi, I see examples of simple CEL validation to ensure that fields for a specific type are not used incorrectly.

That being said, for now, we only support a single type here, which is workload of type `Deployment` correct?
So the validation etc. can be punted for now

However I think I am missing why this TODO is only in `merge.go` -- assuming it was just the first place you saw the usage of the `oneof` proto?

File: projects/gateway2/api/v1alpha1/kube.go (hunk 12 above)
Comment (EItanya) [reply projects/gateway2/api/v1alpha1/kube.go:11]: These are very very light validations. I do think we should get as much validation in as soon as possible, because backfilling validation is very very hard.

File: install/helm/gloo/crds/gateway.gloo.solo.io_gatewayparameters.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (jenshu) [reply install/helm/gloo/crds/gateway.gloo.solo.io_gatewayparameters.yaml]: removed descriptions ece2d6a

File: projects/gateway2/api/doc.go (hunk 1 above)
Comment (jenshu) [reply projects/gateway2/api/doc.go]: removed

File: projects/gateway2/deployer/merge.go (hunk 5 above)
Comment (jenshu) [reply projects/gateway2/deployer/merge.go:95]: probably the most relevant place for a oneof/enum currently is at the top level where we accept either `kube` or `selfManaged`

File: projects/gateway2/api/v1alpha1/kube.go (hunk 12 above)
Comment (jenshu) [reply projects/gateway2/api/v1alpha1/kube.go]: there's an issue for validation https://github.com/solo-io/solo-projects/issues/6014

in the interest of merging this soon, i think the existing minimal validations are good enough

File: projects/gateway2/deployer/merge.go (hunk 5 above)
Comment (jenshu) [reply projects/gateway2/deployer/merge.go]: added `XValidation` tag for the top-level oneof 79d4321b6589c96a5df8559fd3b1c01f91722935

File: projects/gateway2/deployer/merge.go (hunk 4 above)
Comment (jenshu) [reply projects/gateway2/deployer/merge.go]: updated to use getters everywhere

File: projects/gateway2/generate.go (hunk 6 above)
Comment (jenshu) [reply projects/gateway2/generate.go:14]: @lgadban can we resolve / handle this later?

File: projects/gateway2/api/v1alpha1/kube.go (hunk 13 above)
Comment (jenshu) [reply projects/gateway2/api/v1alpha1/kube.go]: updated readme 6163db5

File: projects/gateway2/api/external/kubernetes/README.md (hunk 10 above)
Comment (jenshu) [reply projects/gateway2/api/external/kubernetes/README.md:1]: incorporated some of this info (updated version) into the api readme 6163db5

File: projects/gateway2/api/README.md [hunk 14]
```diff
@@ -1,19 +1,32 @@
...
-- CRDs generated in [install/helm/gloo/crds](/install/helm/gloo/crds)
+These are the steps required to add a new CRD to be used in the Kubernetes Gateway integration:
 
-### Adding a new CRD
+1. If creating a new API version (e.g. `v1`, `v2alpha1`), create a new directory for the version and create a `doc.go` file with the `// +kubebuilder:object:generate=true` annotation, so that Go types in that directory will be converted into CRDs when codegen is run.
+    - The `groupName` marker specifies the API group name for the generated CRD. 
+2. Create a `_types.go` file in the API version directory. Following [gateway_parameters_types.go](/projects/gateway2/api/v1alpha1/gateway_parameters_types.go) as an example:
+    - Define a struct for the resource (containing the metadata fields, `Spec`, and `Status`)
+        - Tip: For spec fields, try to use pointer values when appropriate, as it makes inheritance easier (allows us to differentiate between zero values and nil).
+        - Define getters for each field, as these are not generated automatically.
```
Comment (jbohanon) [inline projects/gateway2/api/README.md:14]: This is a lot of extra work that we lose getting for free. Can we have an intermediate generation step that makes the getters? 

Otherwise I fear we will do what is shown in the test file in this PR where we just don't make the getters and deref everywhere.

File: projects/gateway2/api/v1alpha1/gateway_parameters_types.go (hunk 11 above)
Comment (sam-heilbron) [reply projects/gateway2/api/v1alpha1/gateway_parameters_types.go:1]: Is this something we need to solve (https://github.com/solo-io/gloo/issues/9569) to consider this feature complete? In the past, the value of publishing these APIs was that users could import just our API and build their own controllers to produce those resources. Since this is a public API already, I don't think following the same pattern that we've done is the past is strictly necessary. Curious how you'd like to move forward

File: projects/gateway2/deployer/merge.go (hunk 5 above)
Comment (EItanya) [reply projects/gateway2/deployer/merge.go]: Awesome!!! Speaking of `XValidation` I think it's important that we put list limits in the API ASAP. It doesn't need to happen in this PR specifically because I don't want to block, but I want to make sure we track it. Backfilling length requirements is really really hard as we've seen

**Review by EItanya — DISMISSED** [review]
These changes LGTM, should wait to merge until we have a passing SP PR

File: projects/gateway2/api/v1alpha1/gateway_parameters_types.go (hunk 11 above)
Comment (jenshu) [reply projects/gateway2/api/v1alpha1/gateway_parameters_types.go:1]: we should definitely think about the solo-apis part, but it seems independent of this kubebuilder change

i've added a [comment](https://github.com/solo-io/gloo/issues/9569#issuecomment-2214686390) to that issue to keep in mind the differences between how we'd need to sync gwparams vs our other apis

File: projects/gateway2/api/README.md (hunk 14 above)
Comment (jenshu) [reply projects/gateway2/api/README.md:14]: as discussed offline - would be nice to autogenerate these getters (if we could find / come up with such a tool), but the k8s types that we're importing don't have getters on them anyway so we just need to be careful with our nil checks everywhere

File: projects/gateway2/deployer/merge.go (hunk 5 above)
Comment (jenshu) [reply projects/gateway2/deployer/merge.go]: filed this for length validation https://github.com/solo-io/solo-projects/issues/6507

**Comment by jenshu** [comment]
solo-projects pr that consumes this https://github.com/solo-io/solo-projects/pull/6505
passing ci except a couple failures that seem unrelated

**Review by EItanya — APPROVED** [review]

**Review by jbohanon — APPROVED** [review]

**Comment by jenshu** [comment]
a
